src/Raven.Server/Documents/Indexes/Persistence/IndexFacetReadOperationBase.cs

```diff
@@ -0,0 +1,177 @@
+
+        public int Count;
+        public bool Any;
+
+        public FacetValues(bool legacy)
+        {
+            _legacy = legacy;
+        }
+
+        public void AddDefault(string range)
+        {
+            Any = true;
+            _values[Default] = new FacetValue {Range = range};
+        }
+
+        public void Add(FacetAggregationField field, string range)
+        {
+            if (_legacy)
+            {
+                if (Any)
+                    return;
+
+                AddDefault(range);
+                return;
+            }
+
+            Any = true;
+            _values[field] = new FacetValue {Range = range, Name = string.IsNullOrWhiteSpace(field.DisplayName) ? field.Name : field.DisplayName};
+        }
+
+        public FacetValue Get(FacetAggregationField field)
+        {
+            if (_legacy)
+                return _values[Default];
+
+            return _values[field];
+        }
+
+        public IEnumerable<FacetValue> GetAll()
+        {
+            return _values.Values;
+        }
+
+        public void IncrementCount(int count)
+        {
+            Count += count;
+
+            foreach (var facetValue in _values)
+                facetValue.Value.Count += count;
+        }
+    }
+
+    protected static void UpdateFacetResults(Dictionary<string, FacetedQueryParser.FacetResult> results, IndexQueryServerSide query,
+        Dictionary<string, Dictionary<string, FacetValues>> facetsByName)
+    {
+        foreach (var result in results)
+        {
+            if (result.Value.Ranges != null && result.Value.Ranges.Count > 0)
+                continue;
+
+            var valuesCount = 0;
+            var valuesSumOfCounts = 0;
+            var values = new List<FacetValue>();
+            List<string> allTerms;
+            if (facetsByName?.TryGetValue(result.Key, out var groups) is null or false || groups == null)
+                continue;
+
+            switch (result.Value.Options.TermSortMode)
+            {
+                case FacetTermSortMode.ValueAsc:
+                    allTerms = new List<string>(groups.OrderBy(x => x.Key).ThenBy(x => x.Value.Count).Select(x => x.Key));
+                    break;
+                case FacetTermSortMode.ValueDesc:
+                    allTerms = new List<string>(groups.OrderByDescending(x => x.Key).ThenBy(x => x.Value.Count).Select(x => x.Key));
+                    break;
+                case FacetTermSortMode.CountAsc:
+                    allTerms = new List<string>(groups.OrderBy(x => x.Value.Count).ThenBy(x => x.Key).Select(x => x.Key));
+                    break;
+                case FacetTermSortMode.CountDesc:
+                    allTerms = new List<string>(groups.OrderByDescending(x => x.Value.Count).ThenBy(x => x.Key).Select(x => x.Key));
+                    break;
+                default:
+                    throw new ArgumentException($"Could not understand '{result.Value.Options.TermSortMode}'");
+            }
```

Comment:
Can we avoid LInq here?

Reply:
Well, this part is taken from Lucene facet reader. I can optimize it, we want to use the optimized version in Lucene reader too?

Reply:
I've spoken with @arekpalinski and we think we will skip the optimization part here and do the benchmarking first.